File: Difficulty: Medium/Longest Substring with K Uniques/longest-substring-with-k-uniques.cpp
class Solution {
  public:
    int longestKSubstr(string &s, int k) {
        int n = s.size();
        unordered_map<int, int> map;
        int start = 0, maxLen = -1;
        
        for(int end=0; end<n; end++) {
            map[s[end]]++;
            
            while(start < n && map.size() > k) {
                map[s[start]]--;
                if(map[s[start]] == 0)
                    map.erase(s[start]);
                start++;
            }
            
            if(map.size() == k)
                maxLen = max(maxLen, end - start + 1);
        }
        
        return maxLen;
    }
};

File: Difficulty: Medium/Longest Substring with K Uniques/longest-substring-with-k-uniques.java
class Solution {
    public int longestKSubstr(String s, int k) {
        HashMap<Character,Integer>map = new HashMap<>();
        int ans=-1;
        int j=0;
        for(int i=0;i<s.length();i++)
        {
            char c=s.charAt(i);
            map.put(c,map.getOrDefault(c,0)+1);
            while(map.size()>k)
            {
                map.put(s.charAt(j),map.getOrDefault(s.charAt(j),0)-1);
                if(map.get(s.charAt(j))==0)
                map.remove(s.charAt(j));
                j++;
            }
            if(map.size()==k)
            {
                ans=Math.max(ans,i-j+1);
            }
        }
        
        return ans;
    }
}